File: changemakers/frappe_changemakers/report/donation_allocation/donation_allocation.js
frappe.query_reports["Donation Allocation"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "donor",
			label: __("Donor"),
			fieldtype: "Link",
			options: "Donor",
		},
		{
			fieldname: "donation",
			label: __("Donation"),
			fieldtype: "Link",
			options: "Donation",
		},
		{
			fieldname: "recipient_type",
			label: __("Recipient Type"),
			fieldtype: "Select",
			options: [
				"",
				"Learning Centre",
				"Student",
				"Beneficiary",
				"Budget",
				"Employee",
			],
			default: "",
		},
		{
			fieldname: "beneficiary",
			label: __("Beneficiary"),
			fieldtype: "Link",
			options: "Beneficiary",
			depends_on: 'eval:doc.recipient_type=="Beneficiary"',
		},
		{
			fieldname: "student",
			label: __("Student"),
			fieldtype: "Link",
			options: "Student",
			depends_on: 'eval:doc.recipient_type=="Student"',
		},
		{
			fieldname: "learning_centre",
			label: __("Learning Centre"),
			fieldtype: "Link",
			options: "Learning Centre",
			depends_on: 'eval:doc.recipient_type=="Learning Centre"',
		},
		{
			fieldname: "budget",
			label: __("Budget"),
			fieldtype: "Link",
			options: "Budget",
			depends_on: 'eval:doc.recipient_type=="Budget"',
		},
		{
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",
			depends_on: 'eval:doc.recipient_type=="Employee"',
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "show_total_pledged",
			label: __("Show Pledged Amount"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "show_total_paid",
			label: __("Show Total Paid Amount"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "show_total_distributed",
			label: __("Show Distributed Amount"),
			fieldtype: "Check",
			default: 0,
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "percentage") {
			const percent = Math.min(
				Math.max(data[column.fieldname] || 0, 0),
				100
			);
			const color = "#2ecc71";
			const textInside = percent > 40;
			const percentStr = percent.toFixed(2);

			return `
                <div style="width: 100%; background-color: #e0e0e0; border-radius: 8px; overflow: hidden; position: relative; height: 20px;">
                    <div style="
                        width: ${percent}%;
                        background-color: ${color};
                        height: 100%;
                        transition: width 0.3s ease;
                        text-align: ${textInside ? "center" : "right"};
                        color: white;
                        font-size: 12px;
                        line-height: 20px;
                        padding-right: ${textInside ? "0" : "5px"};
                    ">
                        ${textInside ? `${percentStr}%` : ""}
                    </div>
                    ${
						!textInside
							? `<div style="position: absolute; right: 5px; top: 0; font-size: 12px; line-height: 20px; color: #333;">${percentStr}%</div>`
							: ""
					}
                </div>`;
		}

		if (column.fieldname === "donor" && data && data.donor) {
			return `<a href="/app/donor/${data.donor}" target="_blank">${
				data.donor_name || data.donor
			}</a>`;
		}

		if (
			column.fieldname === "recipient" &&
			data &&
			data.recipient &&
			data.recipient_type
		) {
			const doctype_name = data.recipient_type.replace(/\s/g, "-");
			return `<a href="/app/${doctype_name}/${data.recipient}" target="_blank">${value}</a>`;
		}

		if (column.fieldname === "project" && data && data.project) {
			return `<a href="/app/project/${data.project}" target="_blank">${value}</a>`;
		}

		const blankFields = [
			"pledged_amount",
			"amount",
			"total_amount",
			"total_amount_paid",
			"amount_distributed",
		];

		if (
			blankFields.includes(column.fieldname) &&
			data &&
			(data[column.fieldname] === "" || data[column.fieldname] === "0")
		) {
			return " ";
		}

		return value;
	},
};

File: changemakers/frappe_changemakers/report/donation_allocation/donation_allocation.py
# Copyright (c) 2025, hussain@frappe.io and contributors
# For license information, please see license.txt

import frappe
from collections import defaultdict

def execute(filters=None):
    report = DonationDistributionReport(filters)
    return report.run()

class DonationDistributionReport:
    def __init__(self, filters=None):
        self.filters = frappe._dict(filters or {})
        self.data = []
        self.columns = []
        self.chart = None
        self.report_summary = []

    def run(self):
        self.columns = self.get_columns()
        self.data = self.get_data()
        self.get_chart_data()
        self.get_report_summary()
        return self.columns, self.data, None, self.chart, self.report_summary

    def get_columns(self):
        columns = [
            {"label": "Donation", "fieldname": "donation", "fieldtype": "Link", "options": "Donation", "width": 230},
            {"label": "Donor", "fieldname": "donor", "fieldtype": "Link", "options": "Donor", "width": 150},
        ]

        if self.filters.get("show_total_pledged"):
            columns.append({"label": "Pledged Amount", "fieldname": "pledged_amount", "fieldtype": "Currency", "width": 150})

        if self.filters.get("show_total_paid"):
            columns.append({"label": "Total Paid Amount", "fieldname": "total_amount_paid", "fieldtype": "Currency", "width": 150})

        if self.filters.get("show_total_distributed"):
            columns.append({"label": "Total Distributed Amount", "fieldname": "amount_distributed", "fieldtype": "Currency", "width": 170})

        columns += [
            {"label": "Distribution", "fieldname": "distribution", "fieldtype": "Link", "options": "Donation Allocation", "width": 150},
            {"label": "Date", "fieldname": "distribution_date", "fieldtype": "Date", "width": 150},
            {"label": "Total Amount", "fieldname": "total_amount", "fieldtype": "Currency", "width": 150},
            {"label": "Amount", "fieldname": "amount", "fieldtype": "Currency", "width": 150},
            {"label": "Percentage", "fieldname": "percentage", "fieldtype": "Percent", "width": 150},
            {"label": "Recipient Type", "fieldname": "recipient_type", "fieldtype": "Data", "width": 150},
            {"label": "Recipient", "fieldname": "recipient", "fieldtype": "Dynamic Link", "options": "recipient_type", "width": 150},
            {"label": "Project", "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 150},
            {"label": "Remarks", "fieldname": "remarks", "fieldtype": "Text", "width": 200},
        ]

        return columns


    def get_data(self):
        filters = self.filters
        frappe_filters = {}
        item_filters = {}

        if filters.get("from_date"):
            frappe_filters["date"] = [">=", filters.get("from_date")]
        if filters.get("to_date"):
            frappe_filters.setdefault("date", []).extend(["<=", filters.get("to_date")])
        if filters.get("donation"):
            frappe_filters["donation"] = filters.get("donation")
        if filters.get("donor"):
            frappe_filters["donor"] = filters.get("donor")

        # Dynamic Recipient filters
        if filters.get("recipient_type"):
            item_filters["recipient_type"] = filters["recipient_type"]

            recipient_type = filters["recipient_type"]
            recipient_field_map = {
                "Beneficiary": "beneficiary",
                "Student": "student",
                "Learning Centre": "learning_centre",
                "Budget": "budget",
                "Employee": "employee"
            }

            specific_fieldname = recipient_field_map.get(recipient_type)
            if specific_fieldname:
                recipient_value = filters.get(specific_fieldname)
                if recipient_value:
                    item_filters["recipient"] = recipient_value

        if filters.get("project"):
            item_filters["project"] = filters["project"]


        if isinstance(frappe_filters.get("date"), list) and len(frappe_filters["date"]) > 2:
            dates = frappe_filters.pop("date")[1::2]
            frappe_filters["date"] = ["between", dates]

        distributions = frappe.get_all(
            "Donation Allocation",
            filters=frappe_filters,
            fields=["name", "donation", "donor", "date as distribution_date", "total_amount", "remarks"],
            order_by="donation, date ASC"
        )

        if not distributions:
            return []

        distribution_names = [d.name for d in distributions]
        item_filters["parent"] = ["in", distribution_names]

        items = frappe.get_all(
            "Donation Allocation Item",
            filters=item_filters,
            fields=["parent", "amount", "percentage", "recipient_type", "recipient", "project"],
            order_by="parent, idx"
        )

        items_by_distribution = {}
        for item in items:
            items_by_distribution.setdefault(item.parent, []).append(item)

        distributions = [d for d in distributions if d.name in items_by_distribution]
        donation_names = list(set(d.donation for d in distributions))

        donations = []
        if donation_names:
            donations = frappe.get_all(
                "Donation",
                filters={"name": ["in", donation_names]},
                fields=["name", "donor", "amount as pledged_amount", "total_amount_paid", "amount_distributed"],
                order_by="creation DESC"
            )

        donor_names = {}
        donor_ids = list(set([d.donor for d in donations if d.donor]))
        if donor_ids:
            for d in frappe.get_all("Donor", filters={"name": ["in", donor_ids]}, fields=["name", "donor_name"]):
                donor_names[d.name] = d.donor_name

        distributions_by_donation = {}
        for dist in distributions:
            distributions_by_donation.setdefault(dist.donation, []).append(dist)

        donation_map = {d.name: d for d in donations}
        data = []

        for donation_name in distributions_by_donation.keys():
            donation = donation_map.get(donation_name)
            if not donation:
                continue

            donor_name = donor_names.get(donation.donor, "")
            distributions = distributions_by_donation.get(donation.name, [])

            first_dist = distributions[0]
            first_items = items_by_distribution.get(first_dist.name, [])
            first_item = first_items[0] if first_items else None

            row = {
                "donation": donation.name,
                "donor": donation.donor,
                "donor_name": donor_name,
                "pledged_amount": donation.pledged_amount,
                "total_amount_paid": donation.total_amount_paid or 0,
                "amount_distributed": donation.amount_distributed or 0,
                "is_group": 1,
                "bold": 1,
                "indent": 0
            }

            if first_dist:
                row.update({
                    "distribution": first_dist.name,
                    "distribution_date": first_dist.distribution_date,
                    "total_amount": first_dist.total_amount,
                    "remarks": first_dist.remarks
                })
                if first_item:
                    row.update({
                        "amount": first_item.amount,
                        "percentage": first_item.percentage,
                        "recipient_type": first_item.recipient_type,
                        "recipient": first_item.recipient,
                        "project": first_item.project
                    })

            data.append(row)

            for item in first_items[1:]:
                data.append({
                    "donation": "", "donor": "", "donor_name": "", "pledged_amount": "",
                    "total_amount_paid": "", "amount_distributed": "",
                    "distribution": "", "distribution_date": "", "total_amount": "",
                    "amount": item.amount, "percentage": item.percentage,
                    "recipient_type": item.recipient_type, "recipient": item.recipient,
                    "project": item.project, "remarks": "", "indent": 1
                })

            for dist in distributions[1:]:
                items = items_by_distribution.get(dist.name, [])
                if not items:
                    continue

                first_item = items[0]
                dist_row = {
                    "donation": "", "donor": "", "donor_name": "", "pledged_amount": "",
                    "total_amount_paid": "", "amount_distributed": "",
                    "distribution": dist.name, "distribution_date": dist.distribution_date,
                    "total_amount": dist.total_amount, "remarks": dist.remarks,
                    "is_group": 1, "bold": 0, "indent": 1,
                    "amount": first_item.amount, "percentage": first_item.percentage,
                    "recipient_type": first_item.recipient_type, "recipient": first_item.recipient,
                    "project": first_item.project
                }

                data.append(dist_row)

                for item in items[1:]:
                    data.append({
                        "donation": "", "donor": "", "donor_name": "", "pledged_amount": "",
                        "total_amount_paid": "", "amount_distributed": "",
                        "distribution": "", "distribution_date": "", "total_amount": "",
                        "amount": item.amount, "percentage": item.percentage,
                        "recipient_type": item.recipient_type, "recipient": item.recipient,
                        "project": item.project, "remarks": "", "indent": 2
                    })

        return data


    def get_chart_data(self):

        summary_map = defaultdict(lambda: {"pledged": 0, "paid": 0, "distributed": 0})


        for row in self.data:
            donation = row.get("donation")
            if not donation:
                continue

            indent = row.get("indent", 0)

            if indent == 0:
                summary_map[donation]["pledged"] = row.get("pledged_amount", 0)
                summary_map[donation]["paid"] = row.get("total_amount_paid", 0)
                summary_map[donation]["distributed"] = row.get("amount_distributed", 0)

        items = list(summary_map.items())

        sorted_data = items

        self.chart = {
            "data": {
                "labels": [d[0] for d in sorted_data],
                    "datasets": [
                    {
                    "name": "Pledged Amount",
                    "values": [d[1]["pledged"] for d in sorted_data]
                    },
                    {
                    "name": "Paid Amount",
                    "values": [d[1]["paid"] for d in sorted_data]
                    },
                    {
                    "name": "Distributed Amount",
                    "values": [d[1]["distributed"] for d in sorted_data]
                    }
                    ]
                    },
                    "type": "bar",
                    "fieldtype": "Currency",
                    "height": 300
                }


    def get_report_summary(self):
        total_pledged = 0
        total_paid = 0
        total_distributed = 0

        for row in self.data:
            if row.get("indent", 0) == 0:
                total_pledged += row.get("pledged_amount") or 0
                total_paid += row.get("total_amount_paid") or 0
                total_distributed += row.get("amount_distributed") or 0

        self.report_summary = [
            {
                "label": "Total Pledged Amount",
                "value": total_pledged,
                "datatype": "Currency",
                "indicator": "blue"
            },
            {
                "label": "Total Paid Amount",
                "value": total_paid,
                "datatype": "Currency",
                "indicator": "green"
            },
            {
                "label": "Total Distributed Amount",
                "value": total_distributed,
                "datatype": "Currency",
                "indicator": "orange"
            }
        ]